File: RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd5in79b.py
# *****************************************************************************
# * | File        :	  epd5in79b.py
# * | Author      :   Waveshare team
# * | Function    :   Electronic paper driver
# * | Info        :
# *----------------
# * | This version:   V1.0
# * | Date        :   2024-03-05
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#


import logging
from . import epdconfig

# Display resolution
EPD_WIDTH       = 792
EPD_HEIGHT      = 272

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT

    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.reset_pin, 0)
        epdconfig.delay_ms(1)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200)   

    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)

    # send a lot of data   
    def send_data2(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte2(data)
        epdconfig.digital_write(self.cs_pin, 1)
        
    def ReadBusy(self):
        logger.debug("e-Paper busy")
        while(epdconfig.digital_read(self.busy_pin) == 1):      #  0: idle, 1: busy
            epdconfig.delay_ms(200)
        logger.debug("e-Paper busy release")

    def TurnOnDisplay(self):
        self.send_command(0x22)
        self.send_data(0xF7)            #  24s  #  0xD7  16s  Probability refresh bad, probability damage ink screen
        self.send_command(0x20)			# DISPLAY REFRESH 	
        epdconfig.delay_ms(100)	        # The delay here is necessary, 200uS at least!!!     
        self.ReadBusy()                 # waiting for the electronic paper IC to release the idle signal
            
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
            
        self.reset()

        self.ReadBusy()                 # waiting for the electronic paper IC to release the idle signal
        self.send_command(0x12)         # POWER ON
        self.ReadBusy()                 # waiting for the electronic paper IC to release the idle signal

        self.send_command(0x11)
        self.send_data(0x01)
        self.send_command(0x44)	 		# Set Ram X- address Start / End position
        self.send_data(0x00)     		# XStart, POR = 00h
        self.send_data(0x31)            # 400/8-1
        self.send_command(0x45)	 		# Set Ram Y- address  Start / End position 
        self.send_data(0x0f)  
        self.send_data(0x01)            # 300-1	
        self.send_data(0x00)     		# YEnd L
        self.send_data(0x00)			# YEnd H 

        self.send_command(0x4e)	 						 
        self.send_data(0x00)	
        self.send_command(0x4f)	 
        self.send_data(0x0f)  
        self.send_data(0x01) 	


        self.send_command(0x91)
        self.send_data(0x00)

        self.send_command(0xC4)		    # Set Ram X- address Start / End position
        self.send_data(0x31)     		# XStart, POR = 00h
        self.send_data(0x00)            # 400/8-1
        self.send_command(0xC5)	 	    # Set Ram Y- address  Start / End position 
        self.send_data(0x0f)  
        self.send_data(0x01)            # 300-1	
        self.send_data(0x00)     		# YEnd L
        self.send_data(0x00)			# YEnd H 

        self.send_command(0xCe)	 						 
        self.send_data(0x31)	
        self.send_command(0xCf)	 
        self.send_data(0x0f)  
        self.send_data(0x01)
        return 0

    def getbuffer(self, image):
        # logger.debug("bufsiz = ",int(self.width/8) * self.height)
        buf = [0xFF] * (int(self.width/8) * self.height)
        image_monocolor = image.convert('1')
        imwidth, imheight = image_monocolor.size
        pixels = image_monocolor.load()
        # logger.debug("imwidth = %d, imheight = %d",imwidth,imheight)
        if(imwidth == self.width and imheight == self.height):
            logger.debug("Vertical")
            for y in range(imheight):
                for x in range(imwidth):
                    # Set the bits for the column of pixels at the current position.
                    if pixels[x, y] == 0:
                        buf[int((x + y * self.width) / 8)] &= ~(0x80 >> (x % 8))
        elif(imwidth == self.height and imheight == self.width):
            logger.debug("Horizontal")
            for y in range(imheight):
                for x in range(imwidth):
                    newx = y
                    newy = self.height - x - 1
                    if pixels[x, y] == 0:
                        buf[int((newx + newy*self.width) / 8)] &= ~(0x80 >> (y % 8))
        return buf

    def display(self, imageblack, imagered):
        buf = [0x00] * int(self.width * self.height / 8)
        for i in range(0, int(self.width * self.height / 8)):
            buf[i] = ~imagered[i]

        Width =int(self.width / 16)+1
        Width1 =int(self.width / 8)
        Height = self.height
        
        self.send_command(0x24)
        for i in range(Height):
            self.send_data2(imageblack[i * Width1 : i * Width1+Width])
        self.send_command(0X26)
        for i in range(self.height):
            self.send_data2(buf[i * Width1 : i * Width1+Width])

        self.send_command(0xA4)
        for i in range(self.height):
            self.send_data2(imageblack[i * Width1 + Width - 1 : i * Width1 + Width * 2 - 1])
        self.send_command(0xA6)
        for i in range(self.height):
            self.send_data2(buf[i * Width1 + Width - 1 : i * Width1 + Width * 2 - 1])

        self.TurnOnDisplay()

    def Clear(self):
        self.send_command(0x24)
        self.send_data2([0xFF] * 13600)
        self.send_command(0X26)
        self.send_data2([0x00] * 13600)

        self.send_command(0xA4)
        self.send_data2([0xFF] * 13600)
        self.send_command(0xA6)
        self.send_data2([0x00] * 13600)

        self.TurnOnDisplay()

    def sleep(self):
        self.send_command(0X10) # deep sleep
        self.send_data(0x03)
        
        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

File: Arduino/epd5in79b/epd5in79b.cpp
/**
 *  @filename   :   epd5in79b.h
 *  @brief      :   Header file for e-paper library epd5in83.cpp
 *  @author     :   MyMX from Waveshare
 *  
 *  Copyright (C) Waveshare     2024/03/06
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documnetation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to  whom the Software is
 * furished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdlib.h>
#include "epd5in79b.h"

Epd::~Epd() {
};

Epd::Epd() {
    reset_pin = RST_PIN;
    dc_pin = DC_PIN;
    cs_pin = CS_PIN;
    busy_pin = BUSY_PIN;
    width = EPD_WIDTH;
    height = EPD_HEIGHT;
};

int Epd::Init(void) {
    if (IfInit() != 0) {
        return -1;
    }
    Reset();

    ReadBusy();        //waiting for the electronic paper IC to release the idle signal
	SendCommand(0x12); //POWER ON
	ReadBusy();        //waiting for the electronic paper IC to release the idle signal
	
    SendCommand(0x11);
    SendData(0x01);
    SendCommand(0x44);	 						 // Set Ram X- address Start / End position
	SendData(0x00);     						 // XStart, POR = 00h
	SendData(0x31); //400/8-1
	SendCommand(0x45);	 									// Set Ram Y- address  Start / End position 
	SendData(0x0f);  
	SendData(0x01);  //300-1	
	SendData(0x00);     									// YEnd L
	SendData(0x00);											// YEnd H 

    SendCommand(0x4e);	 						 
	SendData(0x00);	
	SendCommand(0x4f);	 
	SendData(0x0f);  
	SendData(0x01); 	

    
    SendCommand(0x91);
    SendData(0x00);

    SendCommand(0xC4);							 // Set Ram X- address Start / End position
	SendData(0x31);     						 // XStart, POR = 00h
	SendData(0x00); //400/8-1
	SendCommand(0xC5);	 									// Set Ram Y- address  Start / End position 
	SendData(0x0f);  
	SendData(0x01);  //300-1	
	SendData(0x00);     									// YEnd L
	SendData(0x00);											// YEnd H 

    SendCommand(0xCe);	 						 
	SendData(0x31);	
	SendCommand(0xCf);	 
	SendData(0x0f);  
	SendData(0x01);

    return 0;
}

/**
 *  @brief: basic function for sending commands
 */
void Epd::SendCommand(unsigned char command) {
    DigitalWrite(dc_pin, LOW);
    SpiTransfer(command);
}

/**
 *  @brief: basic function for sending data
 */
void Epd::SendData(unsigned char data) {
    DigitalWrite(dc_pin, HIGH);
    SpiTransfer(data);
}

/**
 *  @brief: Wait until the busy_pin goes HIGH
 */
void Epd::ReadBusy(void) {
    while(DigitalRead(busy_pin) == 1) {      //0: busy, 1: idle
        DelayMs(100);
    }      
}

/**
 *  @brief: module reset.
 *          often used to awaken the module in deep sleep,
 *          see Epd::Sleep();
 */
void Epd::Reset(void) {
    DigitalWrite(reset_pin, LOW);                //module reset    
    DelayMs(200);
    DigitalWrite(reset_pin, HIGH);
    DelayMs(200);    
}

/******************************************************************************
function :	Turn On Display
parameter:
******************************************************************************/
void Epd::TurnOnDisplay(void)
{
	SendCommand(0x22); //Display Update Control
	SendData(0xF7);
	SendCommand(0x20); //Activate Display Update Sequence
	ReadBusy();
}

void Epd::Display(const UBYTE *blackimage, const UBYTE *ryimage) {
    int Width, Width1, Height;
    Width =(width % 16 == 0)?(width / 16 ):(width / 16 + 1);
    Width1 =(width % 8 == 0)?(width / 8 ):(width / 8 + 1);
    Height = height;

    // M part 396*272
    SendCommand(0x24);
    for (int j = 0; j < Height; j++) {
        for (int i = 0; i < Width; i++) {
            SendData(blackimage[j * Width1 + i]);
        }
    }

    SendCommand(0X26);
    for (int j = 0; j < Height; j++) {
        for (int i = 0; i < Width; i++) {
            SendData(~ryimage[j * Width1 + i]);
        }
    }

    // S part 396*272
    SendCommand(0xA4);
    for (int j = 0; j < Height; j++) {
        for (int i = 0; i < Width; i++) {
            SendData(blackimage[j * Width1 + i + Width - 1]);
        }
    }

    SendCommand(0XA6);
    for (int j = 0; j < Height; j++) {
        for (int i = 0; i < Width; i++) {
            SendData(~ryimage[j * Width1 + i + Width - 1]);
        }
    }

    TurnOnDisplay();
}

void Epd::Displaypart(const unsigned char* image, unsigned char Block) {
    int Width, Height;
    Width =(width % 16 == 0)?(width / 16 ):(width / 16 + 1);
    Height = height;
    int k = 0;

    if( Block == 0 )
    {
        // M part 396*272
        SendCommand(0x24);
        k = 0;
        for (int j = 0; j < Height; j++) {
            for (int i = 0; i < Width; i++) {
                if(i < 37 && j < 152)
                    SendData((pgm_read_byte(&image[k++])));
                else
                    SendData(0xff);
            }
        }

        SendCommand(0X26);
        for(int i=0; i<13600; i++) 
        {
            SendData(0x00);
        }
    }

    else
    {
        // M part 396*272
        SendCommand(0xA4);
        for(int i=0; i<13600; i++) 
        {
            SendData(0xFF);
        }

        SendCommand(0XA6);
        k = 0;
        for (int j = 0; j < Height; j++) {
            for (int i = 0; i < Width; i++) {
                if(i < 37 && j < 152)
                    SendData(~(pgm_read_byte(&image[k++])));
                else
                    SendData(0x00);
            }
        }

        TurnOnDisplay();
    }
}

void Epd::Clear(void) {
    
    // M part 396*272
    SendCommand(0x24);
    for(int i=0; i<13600; i++) 
    {
        SendData(0xFF);
    }

    SendCommand(0X26);
    for(int i=0; i<13600; i++) 
    {
        SendData(0x00);
    }

    // S part 396*272
    SendCommand(0xA4);
    for(int i=0; i<13600; i++) 
    {
        SendData(0xFF);
    }

    SendCommand(0XA6);
    for(int i=0; i<13600; i++) 
    {
        SendData(0x00);
    }

    TurnOnDisplay();
}


void Epd::Sleep(void) {
    SendCommand(0X10);
    SendData(0x01);
}

/* END OF FILE */


